geo/src/algorithm/bool_ops/safe.rs

```diff
@@ -0,0 +1,467 @@
+use std::f64::consts::PI;
+
+use geo_types::Coord;
+use geo_types::Line;
+use geo_types::LineString;
+use geo_types::Point;
+use geo_types::Polygon;
+
+use crate::BooleanOps;
+use crate::Contains;
+use crate::GeoFloat;
+use crate::MultiPolygon;
+use crate::OpType;
+/// Boolean operations are set operations on geometries considered as a subset
+/// of the 2-D plane. The operations supported are: intersection, union, xor or
+/// symmetric difference, and set-difference on pairs of 2-D geometries and
+/// clipping a 1-D geometry with self.
+///
+/// These operations are implemented on [`Polygon`] and the [`MultiPolygon`]
+/// geometries.
+///
+/// # Validity
+///
+/// Note that the operations are strictly well-defined only on *valid*
+/// geometries. However, the implementation generally works well as long as the
+/// interiors of polygons are contained within their corresponding exteriors.
+///
+/// Degenerate 2-d geoms with 0 area are handled, and ignored by the algorithm.
+/// In particular, taking `union` with an empty geom should remove degeneracies
+/// and fix invalid polygons as long the interior-exterior requirement above is
+/// satisfied.
+pub trait SafeBooleanOps: Sized + BooleanOps {
+    fn preprocess(&self, other: &Self) -> Self;
+
+    fn safe_boolean_op(&self, other: &Self, op: OpType) -> MultiPolygon<Self::Scalar> {
```

Comment:
Yeah that's a nice impulse. I like the idea of going in a checked direction. For me, that would mean that we would probably return an `Option<MultiPolygon<Self::Scalar>>` instead of what we currently have and return `None` in some of the corner cases described in the PR description for which the algo won't work.

That's also closer to [my other attempt of implementing safe boolops](https://github.com/georust/geo/pull/1032). I'll consider that when I'm doing the cleanup

Reply:
I implemented you suggestion and used `Result` as an return type to make it future proof. It actually never fails at the moment but we already have an option to return some errors here and user code will already be in place to respect the case of incoming errors